src/Store/MenuStoreConnector.re

```diff
@@ -9,177 +9,270 @@ module Model = Oni_Model;
+        Isolinear.Effect.none
+      );
+
+    | MenuUpdateSource(source) =>
+      (
+        Option.map((state: Quickmenu.t) => {
+          let count = Quickmenu.getCount(source);
+          {...state, source, selected: min(count, state.selected)}
+        }, state),
+        Isolinear.Effect.none
+      );
+
+    | MenuFocus(index) => (
+      Option.map((state: Quickmenu.t) => {
+        let count = Quickmenu.getCount(state.source);
+
         {
           ...state,
-          selectedItem:
-            position(state.selectedItem, 1, filteredCommandsCount),
-        },
-        Isolinear.Effect.none,
-      )
-    | MenuSearch(query) => (
+          selected: max(0, min(count, index)) // TODO: Could use a clamp function
+        }
+      }, state),
+      Isolinear.Effect.none
+    )
+
+    | NotifyKeyPressed(_, "<UP>")
+    | MenuFocusPrevious => (
+      Option.map((state: Quickmenu.t) => {
+        let count = Quickmenu.getCount(state.source);
+
         {
           ...state,
-          searchQuery: query,
-          filterJob:
-            Core.Job.mapw(MenuJob.updateQuery(query), state.filterJob),
-          selectedItem:
-            position(state.selectedItem, 0, filteredCommandsCount),
-        },
-        queryChangedEffect(state.onQueryChanged, query),
-      )
-    | MenuOpen(menuConstructor) =>
-      let state = Menu.create();
-      (
-        {...state, isOpen: true},
-        menuOpenEffect(menuConstructor, state.onQueryChanged),
-      );
-    | MenuUpdate(update) =>
-      let filterJob =
-        Core.Job.mapw(MenuJob.addItems(update), state.filterJob);
-      let selectedItem =
-        position(state.selectedItem, 0, filteredCommandsCount);
-
-      ({...state, filterJob, selectedItem}, Isolinear.Effect.none);
-    | MenuSetDispose(dispose) => (
-        {...state, dispose},
-        Isolinear.Effect.none,
-      )
-    | MenuClose =>
-      let disposeFunction = state.dispose;
-      (
+          selected:
+            if (count == 0) {
+              0
+            } else if (state.selected <= 0) {
+              count - 1 // "roll over" to end of list
+            } else {
+              state.selected - 1
+            }
+        }
+      }, state),
+      Isolinear.Effect.none
+    )
+
+    | NotifyKeyPressed(_, "<DOWN>")
+    | MenuFocusNext => (
+      Option.map((state: Quickmenu.t) => {
+        let count = Quickmenu.getCount(state.source);
+
         {
           ...state,
-          filterJob: MenuJob.default,
-          isOpen: false,
-          selectedItem: 0,
-          isLoading: false,
-          loadingAnimation: Animation.stop(state.loadingAnimation),
-        },
-        disposeMenuEffect(disposeFunction),
-      );
+          selected:
+            if (count == 0) {
+              0
+            } else {
+              (state.selected + 1) mod count
+            }
+        }
+      }, state),
+      Isolinear.Effect.none
+    )
+
     | MenuSelect =>
```

Comment:
Originally there were multiple different meanings used for the word "select". I've introduced "focus" to indicate the highlighted item in the `Actions.t` type, but haven't followed up on the use elsewhere. It would be easier to just call `MenuSelect` something else, but what...?

Reply:
Hmm, `MenuPick`? `MenuChoose`? One thing we'll have to consider as well is the 'keyboard' focus vs the mouse hover.

Reply:
Yeah, those seem decent. I'll go with one of those. We can revisit later. Disambiguating select lets me revert the use of focus, so that shouldn't be ambiguous then either

Reply:
Hmm, there's a keybinding command too, "menu.select". "menu.choose" certainly doesn't seem as intuitive, and having "menu.select" map to `MenuChoose`, with `MenuSelect` having a different meaning isn't ideal...